DUI3-DX/Connectors/Rhino/Speckle.Connectors.Rhino7/Operations/Receive/RhinoHostObjectBuilder.cs

```diff
@@ -0,0 +1,127 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading;
+using Rhino;
+using Rhino.DocObjects;
+using Rhino.Geometry;
+using Speckle.Connectors.Utils.Builders;
+using Speckle.Converters.Common;
+using Speckle.Core.Models;
+using Speckle.Core.Models.Extensions;
+
+namespace Speckle.Connectors.Rhino7.Operations.Receive;
+
+public class RhinoHostObjectBuilder : IHostObjectBuilder
+{
+  private readonly ISpeckleConverterToHost _toHostConverter;
+  private readonly IConversionContextStack<RhinoDoc, UnitSystem> _contextStack;
+
+  public RhinoHostObjectBuilder(
+    ISpeckleConverterToHost toHostConverter,
+    IConversionContextStack<RhinoDoc, UnitSystem> contextStack
+  )
+  {
+    _toHostConverter = toHostConverter;
+    _contextStack = contextStack;
+  }
+
+  public IEnumerable<string> Build(
+    Base rootObject,
+    string projectName,
+    string modelName,
+    Action<string, double?>? onOperationProgressed,
+    CancellationToken cancellationToken
+  )
+  {
+    var baseLayerName = $"Project {projectName}: Model {modelName}";
+
+    var objectsToConvert = rootObject.TraverseWithPath(obj => obj is not Collection);
+
+    var convertedIds = BakeObjects(objectsToConvert, baseLayerName, onOperationProgressed, cancellationToken);
+
+    _contextStack.Current.Document.Views.Redraw();
+
+    return convertedIds;
+  }
+
+  private List<string> BakeObjects(
+    IEnumerable<(List<string>, Base)> objects,
+    string baseLayerName,
+    Action<string, double?>? onOperationProgressed,
+    CancellationToken cancellationToken
+  )
+  {
+    // LET'S FUCK AROUND AND FIND OUT
```

Comment:
err, we probably should remove that comment, it also doesn't add anything useful?

Reply:
Agreed, it's on me